apps/prairielearn/src/lib/question-render.js

```diff
@@ -17,9 +18,54 @@ import * as questionServers from '../question-servers';
 import * as error from '@prairielearn/error';
 import { getQuestionCourse, ensureVariant } from './question-variant';
 import { writeCourseIssues } from './issues';
+import {
+  AssessmentInstanceSchema,
+  AssessmentSchema,
+  DateFromISOString,
+  GradingJobSchema,
+  IdSchema,
+  IssueSchema,
+  SubmissionSchema,
+  VariantSchema,
+} from './db-types';
 
 const sql = sqldb.loadSqlEquiv(__filename);
 
+const VariantSelectResultSchema = VariantSchema.extend({
+  assessment: AssessmentSchema,
+  assessment_instance: AssessmentInstanceSchema,
+  assessment_instance_date: DateFromISOString,
+  formatted_date: z.string(),
+});
+
+const detailedSubmissionColumns = /** @type {const} */ ({
+  feedback: true,
+  format_errors: true,
+  params: true,
+  partial_scores: true,
+  raw_submitted_answer: true,
+  submitted_answer: true,
+  true_answer: true,
+});
```

Comment:
One neat result of using this approach that one column (client_fingerprint_id) had been omitted from both basic and detailed queries, so this caught the issue.